tensorizer/stream_io.py

```diff
@@ -417,6 +422,292 @@ def seek(self, position, whence=SEEK_SET):
+        return len(self._keys) - 1
+
+    def _read_from(
+        self,
+        position: int,
+        size: int,
+        ba: Union[bytearray, memoryview],
+        no_buffer: bool = False,
+    ) -> int:
+        self.read_operations += 1
+        curr_key = self._find_key_index(position)
+        curr_key_pos = self._indexes[curr_key]
+        curr_key_sz = self._sizes[curr_key]
+        begin_idx = position - curr_key_pos
+        position_end = position + size
+        read_sz = min(size, curr_key_sz)
+        curr_buffer_end = self._curr_buffer_idx + len(self._curr_buffer_view)
+
+        # Check if we have a partial read of the current key in the buffer.
+        # TODO: allow for reads that do not align with the key boundaries.
+        if (
+            not no_buffer
+            and len(self._curr_buffer_view) > 0
+            and self._curr_buffer_idx
+            <= position
+            < position_end
+            < curr_buffer_end
+        ):
+            num_bytes = min(read_sz, len(self._curr_buffer_view))
+            ba[:num_bytes] = self._curr_buffer_view[:num_bytes]
+            self._curr_buffer_idx += num_bytes
+            self._curr_buffer_view = self._curr_buffer_view[num_bytes:]
+            return num_bytes
+
+        self._redis_tcp_mutex.acquire(True)
+
+        # We have an aligned read, so we can use GET.
+        if position == curr_key_pos:
+            command = f"GET {self._keys[curr_key].decode('utf-8')}"
+        else:
+            command = (
+                f"GETRANGE {self._keys[curr_key].decode('utf-8')}"
+                f" {begin_idx} {begin_idx + read_sz - 1}"
+            )
+
+        begin_timer = time.monotonic()
+        self._redis_tcp.send(command.encode("utf-8") + b"\r\n")
+        value_sz = self._read_sz()
+        tcp_read_sz = min(value_sz, read_sz)
+        if tcp_read_sz == 0:
+            self._redis_tcp_mutex.release()
+            return 0
+        num_bytes = self._redis_tcp.recv_into(
+            ba, tcp_read_sz, socket.MSG_WAITALL
+        )
+
+        end_timer = time.monotonic()
+        self.response_latencies.append(end_timer - begin_timer)
+
+        # We performed a partial read of the entire key, so we need to
+        # store the remainder in the buffer.
+        if value_sz > read_sz:
+            tcp_read_sz = value_sz - num_bytes
+            if tcp_read_sz > len(self._curr_buffer):
+                self._curr_buffer = bytearray(tcp_read_sz)
+            self._curr_buffer_view = memoryview(self._curr_buffer)[:tcp_read_sz]
+            self._redis_tcp.recv_into(
+                self._curr_buffer_view, tcp_read_sz, socket.MSG_WAITALL
+            )
+            self._curr_buffer_idx = position + num_bytes
+
+        # Read the trailing \r\n and discard.
+        self._redis_tcp.recv(2)
+
+        self._redis_tcp_mutex.release()
+        return num_bytes
+
+    def _read_sz(self) -> int:
+        # Loop until we get a \r\n
+        sz_resp = b""
```

Comment:
Time complexity, as mentioned in another comment
```suggestion
        sz_resp = bytearray()
```